### DISCUSSION TITLE
Increased response time in recent release

### DISCUSSION THREAD

**Discussion by chuanpan** [open-webui/open-webui#7060] [General]
I recently upgraded from v0.3.35 to v0.4.0 and have noticed that the response time for chat has increased significantly for the same prompt and model. Not sure if other ppl have the same issues?

**Comment by DmitriyAlergant** [comment]
Yes noticed the same, something is sluggish. Not only in chat, admin settings too. General sluggishness. Previous version felt much more responsive...

**Comment by DmitriyAlergant** [comment]
I have a NOT CONFIRMED feeling like if the /models endpoint started being queried more often (?) and as some manifold pipes are querying models from the providers (ex: OpenAI) it is slowing down the UI.   I replaced API call in my OpenAI pipe with a static list, the /models API call improved from 1500ms to 750ms, and the overall UI responsiveness (on changing chats, submitting messages) has noticeably increased - at least it feels that way. 

At this time i cannot positively confirm that the older version was actually different in this aspect.   That being said, some caching of models and less-frequent refresh would definitely be helpful.

**Reply by tjbck** [reply]
Optimization PRs are definitely welcome!

**Reply by smonux** [reply]
Since get_all_models is cached with ttl = 1, would you accept just bumping up to 5-10? It's definitely pretty noticeable right now.

```
@cached(ttl=1)
async def get_all_models():
    models = await get_all_base_models()

    # If there are no models, return an empty list
    if len([model for model in models if not model.get("arena", False)]) == 0:
        return []

    global_action_ids = [
        function.id for function in Functions.get_global_action_functions()
    ]
    enabled_action_ids = [
        function.id
        for function in Functions.get_functions_by_type("action", active_only=True)
    ]
```

**Reply by dolceAlka** [reply]
I've been using a ttl of 90 with a local image, caching models in app.state and it has been working pretty fine, and it is definitely much slower on 0.4. I would recommend a 30 second or above ttl with additional automatic cache refreshes in the main.py by running the refresh every n seconds.

Changing ttl to 180++ and watching logs, I'm not sure cache is even getting hit. response times for get_all_models() doesn't seem to change between caching or not, unsure why for now.

**Reply by tjbck** [reply]
5-10 seconds definitely sounds reasonable, if it significantly improves the latency, PR welcome!

**Reply by dolceAlka** [reply]
I've built it again with higher ttl, and it doesn't seem to help, the issue probably lies elsewhere, its a quick check if anyone wants to verify

**Reply by DmitriyAlergant** [reply]
(incorrect comment)

**Reply by DmitriyAlergant** [reply]
There are internal calls to get_all_models() not triggered explicitly via API from the frontend.

This method is currently being called, synchronously (with an await):
- on all message submissions twice(!) /chat/completions and then once more on /api/chat/completed
- on all tasks (Title Generation, Tags Generation, Query Generation), so first message in the chat is affected even greater

Other users are also noticing get_all_models calls to be the culprit of performance degradation: 
https://github.com/open-webui/open-webui/discussions/7169

This appears have been introduced in the "refac" commit 5 days ago: 
https://github.com/open-webui/open-webui/commit/932de8f1e2d38c93d0c5a9298aa3927800eb77a1#diff-f9b320bb932f19ab661f34fa42616070790451a78da48d2564d38d97f1ee0814R1580

I don't think 5-10s (or even 5-10min) caching will make a significant difference. There must not be a synchronous dependency on get_all_models() for new chat interaction whatsoever.  Models  must be cached globally (across users) and refreshed as background task, plus on admin-initiated configuration updates (Models, Functions, OpenAI, OpenLlama configurations)...

**Reply by DmitriyAlergant** [reply]
@tjbck this may need your attention or at least advice...

**Reply by tjbck** [reply]
Like I said, I'd be happy to accept PRs if there's any tangible differences, I appreciate taking a look!

**Reply by DmitriyAlergant** [reply]
@tjbck it's easier said then done, the code around models catalog was quite convoluted.

Can we begin with a refactoring PR: functions naming de-confusion - there was too many 'get_all_models()" all doing different things, and some logging cleanups? https://github.com/open-webui/open-webui/pull/7199

Merging this at least gets us into position to approach the actual performance-related (caching) fixes...

**Reply by tjbck** [reply]
The issue seems to be resolved with setting the ttl to 3 for me, more testing wanted here since the changes been added to dev images @smonux

**Reply by smonux** [reply]
> The issue seems to be resolved with setting the ttl to 3 for me, more testing wanted here since the changes been added to dev images @smonux

Hi @tjbck , 

I noticed it, not by using the UI, but from calling the API directly while testing my changes. When I merged the 0.4 commits the times went up noticeably.

I have done some unscientific tests, which probably aren't very representative of UI usage 

cache=1s    1m48s
cache=3s    1m42s
cache=10s  1m29s
cache=30s  1m27s
cache=60s  1m29s

Despite the limitations of this "benchmark", I would set it to 10s to be in the safe side.

(I'm using OpenRouter, BTW)

[testtools.py.txt](https://github.com/user-attachments/files/17866824/testtools.py.txt)

**Comment by tjbck** [comment]
<img width="508" alt="image" src="https://github.com/user-attachments/assets/b2f2ba93-3c22-430e-8bc4-20a9cf5703d5">

If you specify the model ids in the connections settings, I can definitely notice the speed improvements. Let me know if this is the case for everyone here.

**Reply by DmitriyAlergant** [reply]
That was the case for me, yes. Though we're integrating through Pipe (built-in OpenAI connection disabled) but it was the same without our manifolds (static hardcoded list of functions performs faster then pulling from the provider). The root cause are multiple unnecessary calls to get_all_models.

**Reply by eureka-0** [reply]
I can confirm that, too.

However, I think we don’t need frequent calls to get_all_models, as the model list changes infrequently. I used a similar web application called LobeChat, and it doesn’t update the model list automatically; instead, we have to update it manually in the settings. This isn’t a significant issue. As for OpenAI, its model list hasn’t changed for several months (gpt-4o-2024-11-20 was added two days ago, replacing the previous version, gpt-4o-2024-08-06).

**Reply by alesha-pro** [reply]
I (and not only me) use Openrouter, and there are 250+ models. Frequent call of get_all_models slows down the application very much.

**Reply by dolceAlka** [reply]
seems like a very good idea, cache until manual refresh, perhaps with a default arg to get_all_models

**Comment by alesha-pro** [comment]
I was just about to post a similar question myself, apparently I'm not the only one with this problem. I've rolled back to 0.3.35 - I'll use that for now until it's fixed.

**Comment by dolceAlka** [comment]
0.4.3 is still significantly slower than 0.3.35 comparing time to first token, but getting models is certainly signficantly faster such as on first load. Not sure why this is as logs seem to indicate that the call to get_all_models is the call taking the time on 0.4.3 vs no call in logs in 0.3.35. Very confusing things. Loading models takes <100ms in network logs but the additional call before sending prompt takes around 1 or 2s in logs.

**Reply by dolceAlka** [reply]
It doesn't seem I was paying attention, there is an additional get_all_models() in apps/openai, so I didn't change that during caching. imo ttl=3 is way too short, especially now having to call get_all_models() between every prompt.

**Comment by smonux** [comment]
I'm going to add bit more of confussion here, but may it's useful for someone.  I have seen this being sent to the LLM (**I don't have web search enabled**).

```
### Task:\\nBased on the chat history, determine whether a search is necessary, and if so, generate a 1-3 broad search queries to retrieve comprehensive and updated information. If no search is required, return an empty list.\\n\\n
### Guidelines:\\n- Respond exclusively with a JSON object.\\n- If a search query is needed, return an object like: { \\"queries\\": [\\"query1\\", \\"query2\\"] } where each query is distinct and concise.\\n- If no search query is necessary, output should be: { \\"queries\\": [] }\\n- Default to suggesting a search query to ensure accurate and updated information, unless it is definitively clear no search is required.\\n- Be concise, focusing strictly on composing search queries with no additional commentary or text.\\n- When in doubt, prefer to suggest a search for comprehensiveness.\\n- Today\'s date is: 2024-11-22\\n\\n
### Output:\\nJSON format: {\\n  \\"queries\\": [\\"query1\\", \\"query2\\"]\\n}\\n\\n### Chat History:\\n<chat_history>\\nUSER: what is the first letter of the latin alphabet\\n</chat_history>\\n"}], "stream": false}'
```

I'm going to follow this .

Edit 1: It doesn't trigger a web search, but a retrieval, but I don't see the point of the llm call. It's inside chat_completion_file_handlers. 
Edit 2: Not sure why this was added, but the queries only should be generated, IMO, if there are files and there's some chance of hitting get_rag_context. There is no point otherwise.
Edit 3: Can anybody try this patch to see if helps?:

```
--- a/backend/open_webui/main.py
+++ b/backend/open_webui/main.py
@@ -811,6 +811,9 @@ async def chat_completion_files_handler(
     contexts = []
     citations = []

+    if not body.get("metadata", {}).get("files", False):
+        return body, {"contexts": contexts, "citations": citations}
+
     try:
         queries_response = await generate_queries(
             {
@@ -836,8 +839,7 @@ async def chat_completion_files_handler(

     print(f"{queries=}")

-    if files := body.get("metadata", {}).get("files", None):
-        contexts, citations = get_rag_context(
+    contexts, citations = get_rag_context(
             files=files,
             queries=queries,
             embedding_function=retrieval_app.state.EMBEDDING_FUNCTION,
@@ -847,7 +849,7 @@ async def chat_completion_files_handler(
             hybrid_search=retrieval_app.state.config.ENABLE_RAG_HYBRID_SEARCH,
         )

-        log.debug(f"rag_contexts: {contexts}, citations: {citations}")
+    log.debug(f"rag_contexts: {contexts}, citations: {citations}")

     return body, {"contexts": contexts, "citations": citations}
```

**Reply by dolceAlka** [reply]
+1, i see this in logs as well

**Reply by smonux** [reply]
It's fixed in 0.4.4!

**Reply by alesha-pro** [reply]
> It's fixed in 0.4.4!

Sure?

**Reply by smonux** [reply]
From the release notes

⏳ RAG Query Generation Issue: Addressed a significant problem where RAG query generation occurred unnecessarily without attached files, drastically improving speed and reducing delays during chat completions.

The commit:
https://github.com/open-webui/open-webui/pull/7262/commits/429fa2befa2ec97b1fc4fffaeefeb4e08fb4287c

**Reply by alesha-pro** [reply]
> From the release notes
> 
> ⏳ RAG Query Generation Issue: Addressed a significant problem where RAG query generation occurred unnecessarily without attached files, drastically improving speed and reducing delays during chat completions.
> 
> The commit: [429fa2b](https://github.com/open-webui/open-webui/commit/429fa2befa2ec97b1fc4fffaeefeb4e08fb4287c)

Im not sure.
For each request openwebui makes at least 2 requests to get_all_models() for each provider. is this normal? Still the latency is quite high compared to 0.3.35
![image](https://github.com/user-attachments/assets/af3317ba-59e7-4233-9ac3-1b6ea06ae4cc)

**Reply by dolceAlka** [reply]
+1 to anonymous

**Comment by chuanpan** [comment]
I just tested it again with the version v0.4.6 and feel the performance has improved!

**Reply by dolceAlka** [reply]
Does seem so
